rust/otap-dataflow/crates/engine/src/exporter.rs

```diff
@@ -16,48 +17,68 @@
 //! 2. The `start` method is called, which begins the exporter's operation
-//! 3. The exporter processes both internal control messages and pipeline data
-//! 4. The exporter shuts down when it receives a `Shutdown` control message or encounters a fatal error
+//! 3. The exporter processes both internal control messages and pipeline data (pdata)
+//! 4. The exporter shuts down when it receives a `Shutdown` control message or encounters a fatal
+//!    error
 //!
 //! # Thread Safety
 //!
-//! Note that this trait uses `#[async_trait(?Send)]`, meaning implementations
-//! are not required to be thread-safe. To ensure scalability, the pipeline engine will start
-//! multiple instances of the same pipeline in parallel, each with its own exporter instance.
+//! Note that this trait uses `#[async_trait(?Send)]`, meaning implementations are not required to
+//! be thread-safe. If you need to implement an exporter that requires `Send`, you can use the
+//! [`SendEffectHandler`] type. The default effect handler is `!Send` (see
+//! [`NotSendEffectHandler`]).
+//!
+//! # Scalability
+//!
+//! To ensure scalability, the pipeline engine will start multiple instances of the same pipeline
+//! in parallel on different cores, each with its own exporter instance.
 
-use crate::NodeName;
+use crate::config::ExporterConfig;
 use crate::error::Error;
 use crate::message::{ControlMsg, Message};
 use async_trait::async_trait;
 use otap_df_channel::error::RecvError;
 use otap_df_channel::mpsc;
+use std::marker::PhantomData;
+use std::pin::Pin;
 use std::rc::Rc;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::time::{Instant, Sleep, sleep_until};
 
 /// A trait for egress exporters.
+///
+/// Note: The default effect handler is `!Send` (see [`NotSendEffectHandler`]).
 #[async_trait(?Send)]
-pub trait Exporter {
-    /// The type of messages handled by the exporter.
-    type PData;
-
+pub trait Exporter<PData, EF = NotSendEffectHandler<PData>>
+where
+    EF: EffectHandlerTrait<PData>,
+{
     /// Starts the exporter and begins exporting incoming data.
     ///
     /// The pipeline engine will call this function to start the exporter in a separate task.
     /// Exporters are assigned their own dedicated task at pipeline initialization because their
     /// primary function involves interacting with the external world, and the pipeline has no
     /// prior knowledge of when these interactions will occur.
     ///
-    /// The `Box<Self>` signature indicates that when this method is called, the exporter takes
-    /// exclusive ownership of its instance. This approach is necessary because an exporter cannot
-    /// yield control back to the pipeline engine - it must independently manage its outputs and
-    /// processing timing. The only way the pipeline engine can interact with the exporter after
-    /// starting it is through the control message channel.
+    /// The exporter is taken as `Box<Self>` so it owns itself once `start` is called.
```

Comment:
> so it owns itself

I think we could be more specific here.

```suggestion
    /// The exporter is taken as `Box<Self>` so the method takes ownership of the exporter once `start` is called.
```